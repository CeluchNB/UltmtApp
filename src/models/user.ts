import { Realm } from '@realm/react'

export const DisplayUserSchema: Realm.ObjectSchema = {
    name: 'DisplayUser',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string',
        localGuest: 'bool?',
    },
}
