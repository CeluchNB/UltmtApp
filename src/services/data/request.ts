import * as Constants from '../../utils/constants'
import { DetailedRequest } from '../../types/request'
import { throwApiError } from '../../utils/service-utils'
import {
    deleteTeamRequest as networkDeleteTeamRequest,
    deleteUserRequest as networkDeleteUserRequest,
    getRequest as networkGetRequest,
    getRequestsByTeam as networkGetRequestsByTeam,
    getRequestsByUser as networkGetRequestsByUser,
    requestTeam as networkRequestTeam,
    requestUser as networkRequestUser,
    respondToPlayerRequest as networkRespondToPlayerRequest,
    respondToTeamRequest as networkRespondToTeamRequest,
} from '../network/request'

/**
 * Method for user to request to join a team.
 * @param token jwt of requesting user
 * @param teamId id of requested team
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const requestTeam = async (
    token: string,
    teamId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRequestTeam(token, teamId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.CREATE_REQUEST_ERROR)
    }
}

/**
 * Method for team to request a user join a team.
 * @param token jwt of team manager
 * @param userId user to request join team
 * @param teamId team requesting the player to join
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const requestUser = async (
    token: string,
    userId: string,
    teamId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRequestUser(token, userId, teamId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.CREATE_REQUEST_ERROR)
    }
}

/**
 * Method to get request details. User must be authorized to view request.
 * User is authorized if 1) he is the user on the request or 2) he is a manager
 * of the team on the request
 * @param token jwt of current user
 * @param requestId id of request
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const getRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkGetRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.GET_REQUEST_ERROR)
    }
}

/**
 * Method for team to respond to a player's request
 * @param token jwt of team manager
 * @param requestId id of request
 * @param accept boolean for accept/deny
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const respondToPlayerRequest = async (
    token: string,
    requestId: string,
    accept: boolean,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRespondToPlayerRequest(
            token,
            requestId,
            accept,
        )
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for a team to delete a pending request it sent.
 * @param token jwt of team manager
 * @param requestId id of request to delete
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const deleteTeamRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkDeleteTeamRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for user to respond to a request from a team
 * @param token jwt of user
 * @param requestId id of request to respond to
 * @param accept boolean for accept/deny
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const respondToTeamRequest = async (
    token: string,
    requestId: string,
    accept: boolean,
): Promise<DetailedRequest> => {
    try {
        const response = await networkRespondToTeamRequest(
            token,
            requestId,
            accept,
        )
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Method for a user to delete a pending request she has sent.
 * @param token jwt of user
 * @param requestId id of request to delete
 * @returns a detailed request object
 * @throws error if backend returns an error
 */
export const deleteUserRequest = async (
    token: string,
    requestId: string,
): Promise<DetailedRequest> => {
    try {
        const response = await networkDeleteUserRequest(token, requestId)
        const { request } = response.data
        return request
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Get all requests belonging to a team
 * @param token token of team manager
 * @param teamId id of team to get requests from
 * @returns list of detailed requests
 */
export const getRequestsByTeam = async (
    token: string,
    teamId: string,
): Promise<DetailedRequest[]> => {
    try {
        const response = await networkGetRequestsByTeam(token, teamId)
        const { requests } = response.data
        return requests
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}

/**
 * Get all requests beloinging to a user
 * @param token token of user to get requests for
 * @returns list of detailed requests
 */
export const getRequestsByUser = async (
    token: string,
): Promise<DetailedRequest[]> => {
    try {
        const response = await networkGetRequestsByUser(token)
        const { requests } = response.data
        return requests
    } catch (error) {
        return throwApiError(error, Constants.REQUEST_RESPONSE_ERROR)
    }
}
