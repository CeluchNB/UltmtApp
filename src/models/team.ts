import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { LocalUser, Team } from '../types/team'

export const DisplayTeamSchema: Realm.ObjectSchema = {
    name: 'DisplayTeam',
    embedded: true,
    properties: {
        _id: 'string',
        place: 'string',
        name: 'string',
        teamname: 'string',
        seasonStart: 'string',
        seasonEnd: 'string',
    },
}

export const GuestTeamSchema: Realm.ObjectSchema = {
    name: 'GuestTeam',
    embedded: true,
    properties: {
        _id: 'string?',
        place: 'string?',
        name: 'string',
        teamname: 'string?',
        seasonStart: 'string?',
        seasonEnd: 'string?',
    },
}

export class TeamSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Team',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            place: 'string',
            name: 'string',
            teamname: 'string',
            managers: { type: 'list', objectType: 'DisplayUser' },
            players: { type: 'list', objectType: 'DisplayUser' },
            seasonStart: 'string',
            seasonEnd: 'string',
            seasonNumber: 'int',
            continuationId: 'string',
            rosterOpen: 'bool',
        },
    }

    _id: string
    place: string
    name: string
    teamname: string
    managers: DisplayUser[]
    players: LocalUser[]
    seasonStart: string
    seasonEnd: string
    seasonNumber: number
    continuationId: string
    rosterOpen: boolean

    constructor(team: Team) {
        this._id = team._id
        this.place = team.place
        this.name = team.name
        this.teamname = team.teamname
        this.managers = team.managers
        this.players = team.players
        this.seasonStart = team.seasonStart
        this.seasonEnd = team.seasonEnd
        this.seasonNumber = team.seasonNumber
        this.continuationId = team.continuationId
        this.rosterOpen = team.rosterOpen
    }
}
