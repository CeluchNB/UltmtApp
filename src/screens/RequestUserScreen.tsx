import * as React from 'react'
import * as RequestData from '../services/data/request'
import * as TeamData from '../services/data/team'
import { DisplayUser } from '../types/user'
import { RequestType } from '../types/request'
import { RequestUserProps } from '../types/navigation'
import ScreenTitle from '../components/atoms/ScreenTitle'
// import SecondaryButton from '../components/atoms/SecondaryButton'
import UserSearchResultItem from '../components/atoms/UserSearchResultItem'
import { searchUsers } from '../services/data/user'
import { selectTeam } from '../store/reducers/features/team/managedTeamReducer'
import { selectToken } from '../store/reducers/features/account/accountReducer'
import { useColors } from '../hooks'
import { useSelector } from 'react-redux'
import { FlatList, StyleSheet, Text, View } from 'react-native'
import { IconButton, TextInput } from 'react-native-paper'
import { size, weight } from '../theme/fonts'

const RequestUserScreen: React.FC<RequestUserProps> = ({ route }) => {
    const { type } = route.params
    const { colors } = useColors()
    const team = useSelector(selectTeam)
    const token = useSelector(selectToken)
    const [players, setPlayers] = React.useState<DisplayUser[]>([])
    const [selectedPlayers, setSelectedPlayers] = React.useState<DisplayUser[]>(
        [],
    )
    const [requestLoading, setRequestLoading] = React.useState(false)
    const [selectedId, setSelectedId] = React.useState('')
    const [error, setError] = React.useState('')
    const [searchError, setSearchError] = React.useState('')

    const search = async (term: string) => {
        setError('')
        if (term.length < 3) {
            setSearchError('')
            setPlayers([])
            return
        }

        try {
            const users = await searchUsers(term)
            if (users.length <= 0) {
                throw new Error()
            }
            setPlayers(users)
        } catch (e) {
            setPlayers([])
            setSearchError('No results from this search, please try again')
        }
    }

    const requestUser = async (user: DisplayUser) => {
        try {
            setError('')
            setRequestLoading(true)
            setSelectedId(user._id)
            switch (type) {
                case RequestType.PLAYER:
                    await RequestData.requestUser(
                        token,
                        user._id,
                        team?._id || '',
                    )
                    break
                case RequestType.MANAGER:
                    await TeamData.addManager(token, team?._id || '', user._id)
                    break
            }
            setSelectedPlayers([user, ...selectedPlayers])
        } catch (e: any) {
            setError(e.message)
        } finally {
            setRequestLoading(false)
        }
    }

    const styles = StyleSheet.create({
        screen: {
            height: '100%',
            backgroundColor: colors.primary,
        },
        title: {
            alignSelf: 'center',
        },
        input: {
            backgroundColor: colors.primary,
            color: colors.textPrimary,
            width: '75%',
            alignSelf: 'center',
            marginBottom: 5,
        },
        error: {
            color: colors.gray,
            width: '75%',
            fontSize: size.fontLarge,
            fontWeight: weight.normal,
            alignSelf: 'center',
        },
        list: {
            width: '75%',
            alignSelf: 'center',
            flexGrow: 0,
        },
        separator: {
            height: 2,
            width: '75%',
            backgroundColor: colors.gray,
            alignSelf: 'center',
            marginBottom: 10,
            marginTop: 10,
        },
        playerContainer: {
            flexDirection: 'row',
        },
        playerItem: {
            color: colors.success,
            fontSize: size.fontMedium,
            fontWeight: weight.full,
            flex: 1,
            alignSelf: 'center',
        },
    })

    return (
        <View style={styles.screen}>
            <ScreenTitle
                style={styles.title}
                title={`Add ${
                    type === RequestType.PLAYER ? 'Players' : 'Managers'
                }`}
            />
            <TextInput
                mode="flat"
                style={[styles.input]}
                underlineColor={colors.textPrimary}
                activeUnderlineColor={colors.textPrimary}
                placeholderTextColor={colors.gray}
                onChangeText={search}
                theme={{
                    colors: {
                        text: colors.textPrimary,
                    },
                }}
                placeholder={`Search ${
                    type === RequestType.PLAYER ? 'players' : 'managers'
                }...`}
            />
            {searchError.length > 0 && (
                <Text style={styles.error}>{searchError}</Text>
            )}
            {/* <SecondaryButton
                text="create bulk join code"
                onPress={async () => {

                }}
            /> */}
            <FlatList
                style={styles.list}
                data={players}
                keyExtractor={item => item._id}
                renderItem={({ item }: { item: DisplayUser }) => (
                    <UserSearchResultItem
                        name={`${item.firstName} ${item.lastName}`}
                        username={`@${item.username}`}
                        loading={selectedId === item._id && requestLoading}
                        onPress={() => requestUser(item)}
                        error={
                            selectedId === item._id && error.length > 0
                                ? error
                                : undefined
                        }
                    />
                )}
            />
            {selectedPlayers.length > 0 && <View style={styles.separator} />}
            <FlatList
                style={styles.list}
                data={selectedPlayers}
                keyExtractor={item => item._id}
                renderItem={({ item }: { item: DisplayUser }) => {
                    return (
                        <View style={styles.playerContainer}>
                            <Text
                                style={
                                    styles.playerItem
                                }>{`${item.firstName} ${item.lastName}`}</Text>
                            <IconButton
                                icon="check"
                                color={colors.success}
                                disabled
                            />
                        </View>
                    )
                }}
            />
        </View>
    )
}

export default RequestUserScreen
