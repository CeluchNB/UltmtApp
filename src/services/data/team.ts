import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import { throwApiError } from '../../utils/service-utils'
import { CreateTeam, Team } from '../../types/team'
import {
    createTeam as networkCreateTeam,
    getManagedTeam as networkGetManagedTeam,
    getTeam as networkGetTeam,
    removePlayer as networkRemovePlayer,
    rollover as networkRollover,
    searchTeam as networkSearchTeam,
    toggleRosterStatus as networkToggleRosterStatus,
} from '../network/team'

/**
 * Method to create team
 * @param token token of user creating team
 * @param data create team data
 * @returns created team
 * @throws error if team is not created successfully
 */
export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<Team> => {
    try {
        const response = await networkCreateTeam(token, data)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.CREATE_TEAM_ERROR)
    }
}

/**
 * Method to search for teams by place, name, or username
 * @param term search term
 * @returns list of teams matching term
 * @throws error if backend returns an error
 */
export const searchTeam = async (term: string): Promise<Team[]> => {
    try {
        if (term.length < 3) {
            throw new ApiError('Not enough characters to search.')
        }
        const response = await networkSearchTeam(term)
        return response.data
    } catch (error) {
        return throwApiError(error, Constants.SEARCH_ERROR)
    }
}

/**
 * Method to get team that the current user manages
 * @param token jwt of current user
 * @param id team id
 * @returns team object
 * @throws error if backend returns error
 */
export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<Team> => {
    try {
        const response = await networkGetManagedTeam(token, id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to get a public team
 * @param id id of team to get
 * @returns team object
 * @throws error if backend returns an error
 */
export const getTeam = async (id: string): Promise<Team> => {
    try {
        const response = await networkGetTeam(id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to change a team's roster status
 * @param token jwt of user changing team
 * @param id id of team to update
 * @param open boolean for team's roster status
 * @returns updated team object
 * @throws error if backend returns error
 */
export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<Team> => {
    try {
        const response = await networkToggleRosterStatus(token, id, open)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to remove a player from a team
 * @param token token of current user
 * @param teamId team player is on
 * @param userId id of player to remove
 * @returns updated team object
 * @throws error if backend returns an error
 */
export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<Team> => {
    try {
        const response = await networkRemovePlayer(token, teamId, userId)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to start a new season for a team. After this method is called,
 * the old team is no longer editable.
 * @param token jwt of manager of team
 * @param teamId id of team to move over
 * @param copyPlayers boolean for deciding if current roster should be kept
 * @param seasonStart start year for season
 * @param seasonEnd start end for season
 * @returns newly created team object
 * @throws error if backend returns an error
 */
export const rollover = async (
    token: string,
    teamId: string,
    copyPlayers: boolean,
    seasonStart: string,
    seasonEnd: string,
): Promise<Team> => {
    try {
        const response = await networkRollover(
            token,
            teamId,
            copyPlayers,
            seasonStart,
            seasonEnd,
        )
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}
