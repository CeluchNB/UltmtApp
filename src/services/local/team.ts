import { Realm } from '@realm/react'
import { Team } from '../../types/team'
import { TeamSchema } from '../../models'
import { getRealm } from '../../models/realm'

export const saveTeams = async (teams: Team[]) => {
    const realm = await getRealm()
    realm.write(() => {
        for (const team of teams) {
            realm.create('Team', team, Realm.UpdateMode.Modified)
        }
    })
}

export const getManagingTeams = async (id: string): Promise<Team[]> => {
    const realm = await getRealm()
    const teams = await realm.objects<TeamSchema>('Team')
    return teams
        .filter(t => t.managers.map(m => m._id).includes(id))
        .map(t => parseTeam(t))
}

const parseTeam = (schema: TeamSchema): Team => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            place: schema.place,
            name: schema.name,
            teamname: schema.teamname,
            managers: schema.managers,
            players: schema.players,
            seasonStart: schema.seasonStart,
            seasonEnd: schema.seasonEnd,
            seasonNumber: schema.seasonNumber,
            continuationId: schema.continuationId,
            rosterOpen: schema.rosterOpen,
            requests: [],
            games: [],
        }),
    )
}
