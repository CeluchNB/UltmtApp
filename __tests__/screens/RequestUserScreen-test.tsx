import * as ManagedTeamReducer from '../../src/store/reducers/features/team/managedTeamReducer'
import * as RequestData from '../../src/services/data/request'
import * as TeamData from '../../src/services/data/team'
import * as UserData from '../../src/services/data/user'
import { NavigationContainer } from '@react-navigation/native'
import { Provider } from 'react-redux'
import React from 'react'
import { RequestUserProps } from '../../src/types/navigation'
import RequestUserScreen from '../../src/screens/RequestUserScreen'
import { Team } from '../../src/types/team'
import store from '../../src/store/store'
import { DetailedRequest, RequestType } from '../../src/types/request'
import { act, fireEvent, render, waitFor } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')
jest.mock('react-native/Libraries/EventEmitter/NativeEventEmitter')

const goBack = jest.fn()
const props: RequestUserProps = {
    navigation: {
        goBack,
    } as any,
    route: {
        params: { type: RequestType.PLAYER },
    } as any,
}

beforeAll(() => {
    store.dispatch(
        ManagedTeamReducer.setTeam({
            _id: 'team1',
        } as Team),
    )

    jest.spyOn(UserData, 'searchUsers').mockReturnValue(
        Promise.resolve([
            {
                _id: 'user1',
                firstName: 'first1',
                lastName: 'last1',
                username: 'first1last1',
            },
            {
                _id: 'user2',
                firstName: 'first2',
                lastName: 'last2',
                username: 'first2last2',
            },
        ]),
    )
})

it('should match snapshot', async () => {
    const snapshot = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    ).toJSON()

    expect(snapshot).toMatchSnapshot()
})

it('should display search results', async () => {
    const { getByPlaceholderText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'first')

    await waitFor(() => queryByText('@first1last1'))
    expect(queryByText('@first2last2')).not.toBeNull()
})

it('should not search with less than 3 characters', async () => {
    const { getByPlaceholderText, findByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'fi')

    await expect(findByText('@first1last1')).rejects.toThrow()
})

it('should display search error', async () => {
    jest.spyOn(UserData, 'searchUsers').mockReturnValueOnce(
        Promise.reject({ message: 'search error' }),
    )

    const { getByPlaceholderText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'first')

    await waitFor(() =>
        queryByText('No results from this search, please try again'),
    )
})

it('should display error with no search results', async () => {
    jest.spyOn(UserData, 'searchUsers').mockReturnValueOnce(Promise.resolve([]))

    const { getByPlaceholderText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'first')

    await waitFor(() =>
        queryByText('No results from this search, please try again'),
    )
})

it('should correctly request user join', async () => {
    const spy = jest.spyOn(RequestData, 'requestUser').mockReturnValueOnce(
        Promise.resolve({
            _id: 'request1',
        } as DetailedRequest),
    )
    const { getByPlaceholderText, findByText, getAllByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'first')

    const user1 = await findByText('@first1last1')
    fireEvent.press(user1)
    await act(async () => {})

    const users = getAllByText('first1 last1')
    expect(users).toHaveLength(2)
    expect(spy).toHaveBeenCalledWith('user1', 'team1')
})

it('should handle request error', async () => {
    const spy = jest.spyOn(RequestData, 'requestUser').mockReturnValueOnce(
        Promise.reject({
            message: 'request error',
        }),
    )
    const { getByPlaceholderText, findByText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search players...')
    fireEvent.changeText(searchInput, 'first')

    const user1 = await findByText('@first1last1')
    fireEvent.press(user1)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('user1', 'team1')
    expect(queryByText('request error')).not.toBeNull()
})

it('should correctly add manager', async () => {
    const spy = jest.spyOn(TeamData, 'addManager').mockReturnValueOnce(
        Promise.resolve({
            _id: 'team1',
        } as Team),
    )

    const managerProps: RequestUserProps = {
        navigation: {
            goBack,
        } as any,
        route: {
            params: { type: RequestType.MANAGER },
        } as any,
    }

    const { getByPlaceholderText, findByText, getAllByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...managerProps} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search managers...')
    fireEvent.changeText(searchInput, 'first')

    const user1 = await findByText('@first1last1')
    fireEvent.press(user1)
    await act(async () => {})

    const users = getAllByText('first1 last1')
    expect(users).toHaveLength(2)
    expect(spy).toHaveBeenCalledWith('team1', 'user1')
})

it('should handle add manager failure', async () => {
    const spy = jest.spyOn(TeamData, 'addManager').mockReturnValueOnce(
        Promise.reject({
            message: 'add manager error',
        }),
    )

    const managerProps: RequestUserProps = {
        navigation: {
            goBack,
        } as any,
        route: {
            params: { type: RequestType.MANAGER },
        } as any,
    }

    const { getByPlaceholderText, findByText, queryByText } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...managerProps} />
            </NavigationContainer>
        </Provider>,
    )

    const searchInput = getByPlaceholderText('Search managers...')
    fireEvent.changeText(searchInput, 'first')

    const user1 = await findByText('@first1last1')
    fireEvent.press(user1)
    await act(async () => {})

    expect(spy).toHaveBeenCalledWith('team1', 'user1')
    expect(queryByText('add manager error')).not.toBeNull()
})

it('should get bulk code', async () => {
    const spy = jest
        .spyOn(TeamData, 'createBulkJoinCode')
        .mockReturnValueOnce(Promise.resolve('123456'))

    const { getByText, getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('create bulk join code')
    fireEvent.press(button)

    expect(spy).toHaveBeenCalled()

    const modal = getByTestId('base-modal')
    await act(async () => {})
    expect(modal.props.visible).toBe(true)
})

it('should handle get bulk code error', async () => {
    const spy = jest
        .spyOn(TeamData, 'createBulkJoinCode')
        .mockReturnValueOnce(Promise.reject({ message: 'bulk code error' }))

    const { getByText, getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('create bulk join code')
    fireEvent.press(button)

    expect(spy).toHaveBeenCalled()

    const modal = getByTestId('base-modal')
    await act(async () => {})
    expect(modal.props.visible).toBe(true)

    const error = getByText('bulk code error')
    expect(error).toBeTruthy()
})

it('should close bulk code modal', async () => {
    const spy = jest
        .spyOn(TeamData, 'createBulkJoinCode')
        .mockReturnValueOnce(Promise.resolve('123456'))

    const { getByText, getByTestId } = render(
        <Provider store={store}>
            <NavigationContainer>
                <RequestUserScreen {...props} />
            </NavigationContainer>
        </Provider>,
    )

    const button = getByText('create bulk join code')
    fireEvent.press(button)

    expect(spy).toHaveBeenCalled()

    const modal = getByTestId('base-modal')
    await act(async () => {})
    expect(modal.props.visible).toBe(true)

    const doneButton = getByText('done')
    fireEvent.press(doneButton)

    const closedModal = getByTestId('base-modal')
    await act(async () => {})
    expect(closedModal.props.visible).toBe(false)
})
