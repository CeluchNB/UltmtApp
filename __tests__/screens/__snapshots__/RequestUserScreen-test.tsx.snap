// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#303030",
      "height": "100%",
    }
  }
>
  <View
    style={
      {
        "alignSelf": "center",
        "display": "flex",
        "flexDirection": "row",
        "width": "90%",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgba(231, 224, 236, 1)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          {
            "backgroundColor": "#303030",
            "color": "#ffd600",
            "flex": 1,
            "marginBottom": 5,
            "width": "75%",
          },
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "#ffd600",
            "bottom": 0,
            "height": 1,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": [
              {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
        testID="text-input-underline"
      />
      <View
        style={
          [
            {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            {
              "minHeight": 56,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          focusable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onPressOut={[Function]}
          placeholder="Search players..."
          placeholderTextColor="#CEC6B4"
          selectionColor="#ffd600"
          style={
            [
              {
                "margin": 0,
              },
              {
                "paddingLeft": 16,
                "paddingRight": 16,
              },
              {
                "height": 56,
              },
              {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              {
                "color": "#ffd600",
                "fontFamily": "System",
                "fontSize": 41.0958904109589,
                "fontWeight": undefined,
                "letterSpacing": 0.15,
                "lineHeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              [
                {},
              ],
              undefined,
            ]
          }
          testID="search-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "alignSelf": "center",
        "bottom": undefined,
        "end": undefined,
        "left": undefined,
        "position": undefined,
        "right": undefined,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
        "start": undefined,
        "top": undefined,
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "transparent",
            "borderColor": "#ffd600",
            "borderRadius": 20,
            "borderStyle": "solid",
            "borderWidth": 1,
            "minWidth": 64,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "overflow": "hidden",
              },
              false,
              {
                "borderRadius": 20,
              },
            ]
          }
          testID="button"
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                [
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "letterSpacing": 0.1,
                    "lineHeight": 20,
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  [
                    {
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    false,
                    {
                      "marginHorizontal": 24,
                      "marginVertical": 10,
                    },
                    undefined,
                    {
                      "textTransform": "uppercase",
                    },
                    {
                      "color": "#ffd600",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "500",
                      "letterSpacing": 0.1,
                      "lineHeight": 20,
                    },
                    undefined,
                  ],
                ]
              }
            >
              create bulk join code
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="base-modal"
    transparent={true}
    visible={false}
  />
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "alignSelf": "center",
        "flexGrow": 0,
        "width": "75%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "alignSelf": "center",
        "flexGrow": 0,
        "width": "75%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
